build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -686,7 +685,8 @@ message Directory {
   repeated SymlinkNode symlinks = 3;
 
   // The node properties of the Directory.
-  repeated NodeProperty node_properties = 4;
+  reserved 4;
```

Comment:
What would be the harm in leaving node_properties as-is, and just adding another field for the field-based ones? That would allow clients that want to be compatible to set both, and servers that want to be compatible to accept both.

Reply:
We would need the same two fields (`repeated NodeProperty` and the new `NodeProperties`) in 5 places in the REAPI and more in downstream protos (e.g., BuildBox uses an additional proto for communication between local components). And the same issue for `NodePropertyKeys`. Code that handles node properties couldn't simply work with a single message. And what name would we choose for the new fields? I.e., it's definitely possible but it makes everything a bit less straight forward long term, in my opinion.

> That would allow clients that want to be compatible to set both, and servers that want to be compatible to accept both.

I don't think we should ever set both variants for the same property (non-canonical, even larger tree size, possibly conflicting redundant values). It would theoretically be possible to use server capabilities to use one or the other approach. I'm not convinced that this would work well in practice, though, as node properties are part of the directory tree, which may also be cached locally. I.e., you might not regenerate the whole tree for a RE action, so you also don't want to tie it to a particular RE server.

Do you already use node properties yourself and are concerned about migration?

Reply:
I just started to work on a Bazel change to specify node properties in some cases to pass additional information to the remote execution server. Technically, if I wait until this change goes in, I could use the new fields instead.

I think what you say makes sense. It would be good to promote some of your reply into the commit message (I think your wording here is clearer than what's there already).

Reply:
I've expanded the commit message.